File: backend/app/services/JWTService.scala
package services

import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import com.auth0.jwt.exceptions.JWTVerificationException
import javax.inject._
import play.api.Configuration
import java.util.{Date, UUID}
import scala.util.{Try, Success, Failure}
import play.api.Logger

case class JWTClaims(userId: UUID, email: String)

@Singleton
class JWTService @Inject()(config: Configuration) {
  
  private val logger = Logger(this.getClass)
  private val secret = config.get[String]("jwt.secret")
  private val expirationTime = config.getOptional[Int]("jwt.expirationHours").getOrElse(24)
  private val algorithm = Algorithm.HMAC256(secret)
  
  def generateToken(userId: UUID, email: String): String = {
    val now = new Date()
    val expiresAt = new Date(now.getTime + (expirationTime * 60 * 60 * 1000))
    
    JWT.create()
      .withIssuer("student-risk-prediction")
      .withSubject(userId.toString)
      .withClaim("email", email)
      .withIssuedAt(now)
      .withExpiresAt(expiresAt)
      .sign(algorithm)
  }
  
  def validateToken(token: String): Option[JWTClaims] = {
    Try {
      val verifier = JWT.require(algorithm)
        .withIssuer("student-risk-prediction")
        .build()
      
      val decodedJWT = verifier.verify(token)
      val userId = UUID.fromString(decodedJWT.getSubject)
      val email = decodedJWT.getClaim("email").asString()
      
      JWTClaims(userId, email)
    } match {
      case Success(claims) => Some(claims)
      case Failure(e: JWTVerificationException) =>
        logger.warn(s"Invalid JWT token: ${e.getMessage}")
        None
      case Failure(e) =>
        logger.error(s"Error validating token: ${e.getMessage}", e)
        None
    }
  }
  
  def refreshToken(oldToken: String): Option[String] = {
    validateToken(oldToken).map { claims =>
      generateToken(claims.userId, claims.email)
    }
  }
}

File: backend-python/app/services/jwt_service.py
import jwt
from datetime import datetime, timedelta
from typing import Optional, Dict
from uuid import UUID
import logging
from app.config import settings

logger = logging.getLogger(__name__)


class JWTService:
    def __init__(self):
        self.secret_key = settings.JWT_SECRET_KEY
        self.algorithm = settings.JWT_ALGORITHM
        self.expiration_hours = settings.JWT_EXPIRATION_HOURS
    
    def generate_token(self, user_id: UUID, email: str) -> str:
        """Generate a JWT token for a user"""
        now = datetime.utcnow()
        expires_at = now + timedelta(hours=self.expiration_hours)
        
        payload = {
            "iss": "student-risk-prediction",
            "sub": str(user_id),
            "email": email,
            "iat": now,
            "exp": expires_at
        }
        
        token = jwt.encode(payload, self.secret_key, algorithm=self.algorithm)
        return token
    
    def validate_token(self, token: str) -> Optional[Dict]:
        """Validate a JWT token and return claims"""
        try:
            payload = jwt.decode(
                token,
                self.secret_key,
                algorithms=[self.algorithm],
                issuer="student-risk-prediction"
            )
            
            return {
                "user_id": UUID(payload["sub"]),
                "email": payload["email"]
            }
        except jwt.ExpiredSignatureError:
            logger.warning("Token has expired")
            return None
        except jwt.InvalidTokenError as e:
            logger.warning(f"Invalid token: {str(e)}")
            return None
        except Exception as e:
            logger.error(f"Error validating token: {str(e)}")
            return None
    
    def refresh_token(self, old_token: str) -> Optional[str]:
        """Refresh a JWT token"""
        claims = self.validate_token(old_token)
        if claims:
            return self.generate_token(claims["user_id"], claims["email"])
        return None


# Singleton instance
jwt_service = JWTService()
